Enable cpu frequency scaling feature on mcxn236

ZhaoxiangJin opened the pull request:
1. Enabled cpu frequency scaling feature on NXP mcxn236
2. Tested with cpu_freq_soc case
```
I: CPU frequency subsystem initialized with interval 1000000 ms
*** Booting Zephyr OS build v4.3.0-rc1-25-g51026e720623 ***
Running TESTSUITE cpu_freq_soc
===================================================================
START - test_soc_pstates
I: 3 P-states defined for frdm_mcxn236/mcxn236
E: P-state is NULL
D: Requesting CPU frequency change: mode_idx=2 -> 150000000 Hz (threshold=60%)
D: Switching CPU clock to 150 MHz
D: Requesting CPU frequency change: mode_idx=1 -> 48000000 Hz (threshold=30%)
D: Switching CPU clock to 48 MHz
D: Requesting CPU frequency change: mode_idx=0 -> 12000000 Hz (threshold=10%)
D: Switching CPU clock to 12 MHz
 PASS - test_soc_pstates in 0.029 seconds
===================================================================
TESTSUITE cpu_freq_soc succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [cpu_freq_soc]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.029 seconds
 - PASS - [cpu_freq_soc.test_soc_pstates] duration = 0.029 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

3. Tested with on_demand policy
```
I: CPU frequency subsystem initialized with interval 1000000 ms
*** Booting Zephyr OS build v4.3.0-rc1-89-g6aa64688b83b ***
Running TESTSUITE cpu_freq_on_demand
===================================================================
START - test_pstates
E: On-Demand Policy: pstate_out is NULL
D: CPU0 Execution cycles: 18832087, Total cycles: 18832087
D: CPU0 Load: 100%
D: On-Demand Policy: Selected P-state 0 with load_threshold=60%
D: CPU0 Execution cycles: 16865960, Total cycles: 1863451
D: CPU0 Load: 11%
D: On-Demand Policy: Selected P-state 2 with load_threshold=0%
 PASS - test_pstates in 0.229 seconds
===================================================================
TESTSUITE cpu_freq_on_demand succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [cpu_freq_on_demand]: pass = 1, fail = 0, skip = 0, total = 1 duration = 0.229 seconds
 - PASS - [cpu_freq_on_demand.test_pstates] duration = 0.229 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION SUCCESSFUL
```

ZhaoxiangJin committed:
boards: frdm_mcxn236: Update CLKOUT source clock

Use the primary clock source FRO12M as the clock source for
CLKOUT. Compared with the secondary clock source(main clock,
the primary clock source is more stable and will not cause
timing problems of camera PCLK, HSYNC and VSYNC due to
unexpected changes.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
dts: bindings: p_state: Add bindings for NXP mcxn

1. Add p-state binding for nxp mcxn series.
2. Add p-state nodes for nxp mcxn series.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: mcxn: enable p-state driver for nxp mcxn

Enable p-state driver for nxp mcxn, currently tested
on frdm-mcxn236 but should be easily extendable to
other mcxn boards.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: cpu_freq: Enable cpu_freq_soc test on frdm_mcxn236

Enable cpu_freq_soc test on frdm_mcxn236

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

seankyer commented:
Question, will your board support the `on_demand` policy? If so, you may consider adding your board to the sample.yaml in https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/cpu_freq/on_demand?

ZhaoxiangJin commented:
> Question, will your board support the `on_demand` policy? If so, you may consider adding your board to the sample.yaml in https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/subsys/cpu_freq/on_demand?

Hello, I will add this case.

ZhaoxiangJin committed:
tests: cpu_freq: Enable on_demand policy test on frdm_mcxn236

Enable on_demand policy test on frdm_mcxn236

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
tests: cpu_freq: on_demand: Increase WAIT_US

A low-load period that is too short (1ms) will cause
the average load to remain almost unchanged. Increase
it to 100ms.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin commented:
Hello @hakehuang, may I request you kick a test for this PR? Thanks!

hakehuang reviewed (dismissed):
some small question. not blocker

ZhaoxiangJin commented on tests/subsys/cpu_freq/policies/on_demand/src/main.c:
```diff
@@ -13,7 +13,7 @@
 LOG_MODULE_REGISTER(cpu_freq_on_demand_test, LOG_LEVEL_INF);
 
-#define WAIT_US 1000
+#define WAIT_US 100000
```
The test case try to sleep 1ms [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/subsys/cpu_freq/policies/on_demand/src/main.c#L53) to create a low-load scenario, but between two CPU load measurement points, operations such as log processing, printing, and assertions will keep the system busy, 1ms will not cause the overall system load to decrease. So need to increase the idle time.
with the 1ms sleep, part of the log is:
```
START - test_pstates
E: On-Demand Policy: pstate_out is NULL
D: CPU0 Execution cycles: 3982062, Total cycles: 3982062
D: CPU0 Load: 100%
D: On-Demand Policy: Selected P-state 0 with load_threshold=60%
D: CPU0 Execution cycles: 2001001, Total cycles: 1837470
D: CPU0 Load: 91%
D: On-Demand Policy: Selected P-state 0 with load_threshold=60%
```
In the low-load scenario of the test case, the CPU load is 91%, which is not significantly different from the high-load scenario. Therefore, I believe that changing the sleep time is a more universal approach (compared to changing the load-threshold in devicetree).

ZhaoxiangJin commented on boards/nxp/frdm_mcxn236/board.c:
```diff
@@ -315,12 +315,12 @@ void board_early_init_hook(void)
 	 */
-	CLOCK_AttachClk(kMAIN_CLK_to_CLKOUT);
-	CLOCK_SetClkDiv(kCLOCK_DivClkOut, 25U);
+	CLOCK_AttachClk(kFRO12M_to_CLKOUT);
```
I recommend using a primary clock source (FRO12M) here because this clock is a stable 12MHz and cannot be accidentally tampered with. If a secondary clock source, such as the main clock, then adjusting the main clock will cause the camera clock to change.
Furthermore, from a design perspective, all peripherals' clock sources on the MCXN236 come from a primary clock source, rather than from the main clock, which provides the basis for CPU frequency scaling feature.

seankyer commented on dts/arm/nxp/nxp_mcxn23x_common.dtsi:
```diff
@@ -30,6 +30,26 @@
+
+		pstate_2: pstate-2 {
+			compatible = "nxp,mcxn23x-pstate";
+			load-threshold = <10>; //No basis, just assumed threshold
```
It should really be better dictated in the documentation, but as you noticed when adding support for the on_demand sample, the lowest p-state should have a load-threshold of 0, or the on_demand policy will get stuck in the current state if load drops below 10%

ZhaoxiangJin replied:
Since there is only on_demand policy at the present and the on_demand logic has shortage, I am agree change the threshold to '0' in soc dtsi level.

seankyer commented on tests/subsys/cpu_freq/policies/on_demand/boards/frdm_mcxn236.overlay:
```diff
@@ -0,0 +1,32 @@
+
+		pstate_2: pstate-2 {
+			status = "okay";
+			/* The on_demand policy selection rule is to sequentially
```
You identified something important about the on_demand policy: if it doesn't find a P-state with a threshold lower than the current load, it will fail to switch state. Which may not be intended if load went from 100% -> 0% suddenly as the chip would get stuck in a "turbo mode" meanwhile its load is low.

Instead of this comment in the overlay, please document this behaviour in the on_demand policy docs.

I also think we should add some logic to the on_demand policy to pick the lowest level if it fails to find one, or add a build_assert to ensure the board has defined its "0 load" (default, I suppose) p-state. We can do this later if you don't want to add that to this PR however.

ZhaoxiangJin replied:
I'm not fixated on whether this comment must exist or not, so if you think it's unnecessary, I can certainly update the PR to remove it.

seankyer replied:
in my opinion, remove it since we're updating the CPUFreq docs and on_demand policy, all thanks to your effort :)

seankyer commented on soc/nxp/mcx/mcxn/cpu_freq_scale.c:
```diff
@@ -0,0 +1,213 @@
+
+	switch (cfg->mode_idx) {
+	case 2U: /* high */
+		target_hz = 150000000U;
```
I suggest you create a #define for these magic numbers (used a few times in this file). Or you could even add it to your custom pstate device tree binding.

seankyer commented on soc/nxp/mcx/mcxn/cpu_freq_scale.c:
```diff
@@ -0,0 +1,213 @@
+	uint32_t target_hz;
+
+	switch (cfg->mode_idx) {
+	case 2U: /* high */
```
You could define an actual enum for these cases, just a suggestion.

seankyer reviewed (approved):
from a CPUFreq perspective LGTM
